python/mlc_chat/compiler/model/position_embedding_op.py

```diff
@@ -0,0 +1,99 @@
+"""Operators for positional embeddings, e.g. RoPE."""
+from typing import Tuple
+from tvm import tir
+from tvm.relax.frontend.nn import Tensor, op
+from tvm.script import tir as T
+
+# pylint: disable=invalid-name
+
+def llama_rope(
+    qkv: Tensor,
+    offset: tir.Var,
+    rope_theta: float,
+    num_q_heads: int,
+    num_kv_heads: int,
+) -> Tuple[Tensor, Tensor, Tensor]:
+    """Llama-style RoPE. Given a fused QKV tensor, it returns three tensors, Q, K, and V, where Q
+    and K are rotated by RoPE while V remains unchanged.
+
+    Parameters
+    ----------
+    qkv : Tensor
+        The fused QKV tensor of shape: [batch_size, seq_len, #q_heads + #kv_heads * 2, head_dim]
+
+    offset : tir.Var
+        The offset applied to `seq_len`. The `s`-th position in qkv, i.e. `qkv[_, s, _, _]`,
+        corresponds to `s + offset` in the full sequence after being concatenated with KVCache.
+
+    rope_theta: float
+        The theta value in RoPE, which controls the frequency.
+
+    num_q_heads : int
+        The number of query heads.
+
+    num_kv_heads : int
+        The number of key/value heads. It differs from `num_q_heads` in group-query attention.
+
+    Returns
+    -------
+    q : Tensor
+        The query tensor of shape [batch_size, seq_len, #q_heads, head_dim] w/ RoPE applied
+
+    k : Tensor
+        The key tensor of shape [batch_size, seq_len, #kv_heads, head_dim] w/ RoPE applied
+
+    v : Tensor
+        The value tensor of shape [batch_size, seq_len, #kv_heads, head_dim] w/o RoPE applied
+    """
+    _, _, fused_heads, head_dim = qkv.shape
+    assert fused_heads == num_q_heads + num_kv_heads * 2
+    dtype = qkv.dtype
+    head_dim = tir.const(head_dim, "int32")
+    rope_theta = tir.const(rope_theta, "float32")
+
+    def _rope( # pylint: disable=too-many-arguments
+        x: T.Buffer,
+        b: tir.Var,
+        s: tir.Var,
+        h: tir.Var,
+        d: tir.Var,
+        offset: tir.Var,
+    ):
+        freq = (offset + s) / rope_theta.power(d * 2 % head_dim / head_dim.astype("float32"))
```

Comment:
Note that `freq` and its `sin`/`cos` only depend on `s` and `d`, and `d` is usually around `[0, 128)` as `head_dim`. It means it's possible to further optimize rope compute by caching `{sin/cos}(freq)` on chip. However, it doesn't seem a significant bottleneck in our testing, so I'm not heavily optimizing it in this PR